db/crud.go

```diff
@@ -939,6 +939,31 @@ func (db *Database) PutExistingRevWithConflictResolution(newDoc *Document, docHi
 			return nil, nil, nil, base.ErrUpdateCancel // No new revisions to add
 		}
 
+		addRevisionsToRevTree := func() error {
+			// Add all the new-to-me revisions to the rev tree:
+			for i := currentRevIndex - 1; i >= 0; i-- {
+				err := doc.History.addRevision(newDoc.ID,
+					RevInfo{
+						ID:      docHistory[i],
+						Parent:  parent,
+						Deleted: i == 0 && newDoc.Deleted})
+
+				if err != nil {
+					return err
+				}
+				parent = docHistory[i]
+			}
+			return nil
+		}
+
+		if doc.IsDeleted() && newDoc.Deleted {
+			err = addRevisionsToRevTree()
```

Comment:
In the case of a remote tombstone, where the local doc is already tombstoned, I didn't think it would be necessary to add the remote tombstone to the rev tree unless it's tombstoning an active leaf revision.  Checking for that case should already be handled inside IsIllegalConflict.  So I think you just need to catch the case where IsIllegalConflict returns true, and both are tombstones (i.e. check for both deleted after line 968 and terminate the update at that point).

I don't think it's necessary to add the remote tombstone to the revtree for the same reason we're not

Reply:
I forgot where we got up to on the CBL P2P + SG-replicate co-existence. Do we not need to replicate all tombstones in sg-replicate to handle clients that have only pulled one side of the tree?

Reply:
As discussed, I think we can just modify IsIllegalConflict to not reject tombstones when being called from PutExistingRevWithConflictResolution (currently it allows tombstones that are tombstoning an active leaf, the change would be to allow all tombstones when running with conflict resolution).

Reply:
To validate this functionality, we should have unit tests for tombstones on local and remote that cover the following:
1. two tombstones, rev local > rev remote 
2. two tombstones, rev remote > rev local
3. For each of the above two cases, add tests for:
     - After replication completes, resurrect local, ensure it is replicated to remote
     - After replication completes, resurrect remote, ensure it is replicated to local

Reply:
Think this should be done now 👍